### DISCUSSION TITLE
Build Inventory in Code

### DISCUSSION THREAD

**Discussion by bgile** [nornir-automation/nornir#738] [General]
I used to do this with Nornir 2 and then once I updated to Nornir 3 I hadn't really bothered but in some cases, it's easier for me to just create an inventory on the fly when things are dynamic rather than have a predefined YAML file.  Currently, I just build the YAML file and store it as a workaround, but I would like to get back to doing this in line with code when it suits me.  Looking to see if anyone else is doing this and what the code might look like.  The working example below is Nornir 2 so core= becomes runner= I belive and such, mostly the first part with SimpleInventory that I'm getting stuck on.  I was attempting something like this:

```python
nr = InitNornir(
    inventory={
        "plugin": "SimpleInventory",
        "options":{
            "hosts":{
                "some host":{
                    "hostname": "some host",
                    "username": "some user",
                    "port": 22,
                    "password": "some password",
                    "platform": "junos",
                    "groups": ["Juniper"]
                }
            },
            "groups": {"Juniper": {}}
        }
    },
    runner={
        "plugin": "threaded",
        "options": {
            "num_workers": 1
        },
    },
    logging={
        "enabled": False
    }
)
```

And getting this exception:

```---> 18     inv = inventory_plugin(**config.inventory.options).load()
     19 
     20     if config.inventory.transform_function:

TypeError: __init__() got an unexpected keyword argument 'hosts'
```

Example of old Nornir 2 code that worked:

```python
nr = InitNornir(inventory={
    "plugin": "nornir.plugins.inventory.simple.SimpleInventory",
    "options": {
        "hosts": {
            device: {
                "hostname": device,
                "port": 22,
                "username": user,
                "password": pw,
                "platform": "junos",
                "groups": ["Juniper"],
            }
        },
        "groups": {"Juniper": {}}
    }
}, core={
    "num_workers": 1
}, logging={
    "enabled": True,
    "level": "debug",
    "file": f"log_files/{device}-{time_now}_data.log",
})
```

**Comment by dbarrosop** [comment]
Creating an inventory from a dictionary isn't currently supported. Shouldn't be hard to create an inventory plugin that takes a dictionary and creates an inventory. Specially after #727 is completed.

If you want to go down that path you may want to contribute it to https://github.com/nornir-automation/nornir_utils as it may be useful to others.

**Reply by bgile** [reply]
Thanks for the feedback @dbarrosop, I will look at this.  I may dedicate some cycles to look into crafting a util, I think there is some value in it.  Cheers.

**Comment by ubaumann** [comment]
I never used it but nornir-salt has a DictInventory Plugin https://nornir-salt.readthedocs.io/en/latest/Inventory%20Plugins.html#dictinventory-plugin

**Reply by bgile** [reply]
Hmm this is interesting.  Thanks for the find, this might actually work with what I want to do, and at the very least be something I can use in different situations.

Update: This worked perfectly for my application.  Thanks again!
